synapse/storage/databases/main/event_federation.py

```diff
@@ -922,6 +962,26 @@ def _get_backfill_events(self, txn, room_id, event_list, limit):
                 if row[1] not in event_results:
                     queue.put((-row[0], row[1]))
 
+            # From each event, go forwards through successors to find chains of
+            # historical backfilled floating outliers
+            successorQueue = PriorityQueue()
+            successorQueue.put(event_id)
+            while not successorQueue.empty() and len(event_results) < limit:
+                try:
+                    event_id = successorQueue.get_nowait()
+                except Empty:
+                    break
+
+                successor_event_ids = self.get_successor_events_txn(txn, [event_id])
+                for successor_event_id in successor_event_ids:
+                    # Skip any branches we have already gone down
+                    if successor_event_id in event_results:
+                        continue
+
+                    event_results.add(successor_event_id)
+
+                    successorQueue.put(successor_event_id)
```

Comment:
*Pulled out of discussion from https://github.com/matrix-org/synapse/pull/9247/files#r634269224*

---

> Shouldn't we be looking at the marker events here? I don't think we want to be looking at the prev events anymore?

In order to make this work, should we add a new `insertion_event_lookups` table to store which events are marked as insertion points? It stores, `insertion_event_id` and  `insertion_prev_event_id`

This way when we're iterating back in time, we can check whether the current `event_id` is `insertion_prev_event_id` in the lookup table. When it matches, we can iterate from `insertion_event_id` by successors up each chunk. Are there performance problems with iterating successors?

![](https://user-images.githubusercontent.com/558581/118685772-fbc0f080-b7c8-11eb-96a1-793ef9222928.png)


---


When a remote federated homeserver, receives a "marker" event, it could

 - Backfill the "insertion" event immediately to add it to the `insertion_event_lookups` table. The rest of the chunk does not backfill yet.
 - Add the "insertion" event to the `event_backward_extremities`. When someone uses scrollback, it will request the historical messages and the navigation strategy above will return the correct messages.

Reply:
To summarize the ideas from [today's sync call](https://docs.google.com/document/d/1JJDuPfcPNX75fprdTWlxlaKjWOdbdJylbpZ03hzo638/edit#bookmark=id.ju4icm9d1kvp) with @erikjohnston and myself:

My previous diagrams were ambiguous in terms of showing which historical messages were the most recent but the discussion was always grounded around reverse-chronological which made them make sense mostly to me since I had the assumption loaded in my mind. I'll make the diagrams more obvious here!


### Expectation

Here is what scrollback is expected to look like in Element:

![](https://user-images.githubusercontent.com/558581/119064795-cae7e380-b9a1-11eb-9366-5e1f5e6370a8.png)


### Current

Currently, the DAG structure looks like the following because the batch insert request payload is in **reverse-chronological** order (`[2, 1, 0]`), we process it in that order which sets the `prev_events` and then we persist it that way to get the correct `(topological_ordering, stream_ordering)` order so it sorts between A and B as we expect.

![](https://user-images.githubusercontent.com/558581/119064864-f074ed00-b9a1-11eb-91a6-bcd6398014d4.png)




### New approach

But what we want to do is have the payload in **chronological** order (`[0, 1, 2]`), process it in that order so the `prev_events` point to it's older previous message which is more sane, and then when persisting, we **reverse the list (to make it reverse-chronological)** so we can still get the correct `(topological_ordering, stream_ordering)` order so it sorts between A and B as we expect.

With the new process, the DAG will look like:

![](https://user-images.githubusercontent.com/558581/119065959-4f3b6600-b9a4-11eb-9e23-2e3769e74679.png)


### New approach with "insertion" events

With "insertion" events, we just add them to the start of each chronological chunk (where the oldest message in the chunk is). The next older in time chunk can connect to that "insertion" point from the previous chunk.


[Mermaid live editor playground link](https://mermaid-js.github.io/mermaid-live-editor/#/edit/eyJjb2RlIjoiZmxvd2NoYXJ0IEJUXG4gICAgc3ViZ3JhcGggbGl2ZVxuICAgICAgICBCIC0tLS0tLS0tLS0tLT4gQVxuICAgIGVuZFxuICAgIFxuICAgIHN1YmdyYXBoIGNodW5rMFxuICAgICAgICBjaHVuazAtMigoXCIyXCIpKSAtLT4gY2h1bmswLTEoKDEpKSAtLT4gY2h1bmswLTAoKDApKSAtLT4gY2h1bmswLWluc2VydGlvblsvaW5zZXJ0aW9uXFxdXG4gICAgZW5kXG5cbiAgICBzdWJncmFwaCBjaHVuazFcbiAgICAgICAgY2h1bmsxLTIoKFwiMlwiKSkgLS0-IGNodW5rMS0xKCgxKSkgLS0-IGNodW5rMS0wKCgwKSkgLS0-IGNodW5rMS1pbnNlcnRpb25bL2luc2VydGlvblxcXVxuICAgIGVuZFxuICAgIFxuICAgIHN1YmdyYXBoIGNodW5rMlxuICAgICAgICBjaHVuazItMigoXCIyXCIpKSAtLT4gY2h1bmsyLTEoKDEpKSAtLT4gY2h1bmsyLTAoKDApKSAtLT4gY2h1bmsyLWluc2VydGlvblsvaW5zZXJ0aW9uXFxdXG4gICAgZW5kXG5cbiAgICBcbiAgICBjaHVuazAtaW5zZXJ0aW9uIC0tLS0tLS0-IEFcbiAgICBjaHVuazEtaW5zZXJ0aW9uIC0tPiBBXG4gICAgY2h1bmsxLTIgLS4tPiBjaHVuazAtaW5zZXJ0aW9uXG4gICAgY2h1bmsyLWluc2VydGlvbiAtLT4gQVxuICAgIGNodW5rMi0yIC0uLT4gY2h1bmsxLWluc2VydGlvbiIsIm1lcm1haWQiOnt9LCJ1cGRhdGVFZGl0b3IiOmZhbHNlfQ)
<details>
<summary>mermaid graph syntax</summary>

```mermaid
flowchart BT
    subgraph live
        B ------------> A
    end
    
    subgraph chunk0
        chunk0-2(("2")) --> chunk0-1((1)) --> chunk0-0((0)) --> chunk0-insertion[/insertion\]
    end

    subgraph chunk1
        chunk1-2(("2")) --> chunk1-1((1)) --> chunk1-0((0)) --> chunk1-insertion[/insertion\]
    end
    
    subgraph chunk2
        chunk2-2(("2")) --> chunk2-1((1)) --> chunk2-0((0)) --> chunk2-insertion[/insertion\]
    end

    
    chunk0-insertion -------> A
    chunk1-insertion --> A
    chunk1-2 -.-> chunk0-insertion
    chunk2-insertion --> A
    chunk2-2 -.-> chunk1-insertion
```

</details>

![](https://user-images.githubusercontent.com/558581/119069613-130c0380-b9ac-11eb-82a8-ed67ada77d71.png)


### "Marker" events

  - > we add a new `insertion_event_lookups` table to store which events are marked as insertion points? It stores, `insertion_event_id` and `insertion_prev_event_id`, https://github.com/matrix-org/synapse/pull/9247/files#r636520393
 - We don't need to add a "marker" event for each chunk. We can batch insert multiple chunks then once we're done importing everything, we can add one "marker" event pointing at the root "insertion" point
    - For the purposes of initial MVP for Beeper, we don't need to add "marker" events at all
    - The "marker" events can always be added later if necessary (for federation or whatever) and is the beauty of having them!
 - When a remote federated homeserver, receives a "marker" event, we just add the details to the `insertion_event_lookups` table (no fetching or backfilling the "insertion" event)
    - We discussed not wanting to fetch the "insertion" event when the "marker" comes down the pipe but I've realized that in order to store `insertion_prev_event_id` in the table, we either need to a) add it as part of the "marker" event which works to not fetch anything additional or b) backfill just the "insertion" event to get it. I think I am going to opt for option A though

So the structure of the marker event would look like:
```js
{
    "type": EventTypes.MSC2716_MARKER,
    "sender": "xxx",
    "content": {
        EventContentFields.MSC2716_INSERTION: insertion_event.event_id,
        EventContentFields.MSC2716_INSERTION_PREV_EVENTS: insertion_event.prev_events,
    },
    "room_id": room_id,
}
```

We need to be able to lookup a given event against `insertion_prev_event_id` in the lookup table. Since `prev_events` is an array, I guess we insert an entry in `insertion_event_lookups` for each `prev_event`. Or if there is a performant index for array fields, that's seems fine to store `insertion_prev_event_ids`(plural).

Reply:
I got the new approach working 🎉  with a small caveat around needing to comment out `action_generator` in `synapse/handlers/message.py` (see todo comment there).

<img width="300" alt="Room in Element with backfilled messages" src="https://user-images.githubusercontent.com/558581/119115004-4e2e2700-b9ec-11eb-8ba8-7d26ff258dcb.png">

The DAG also looks sane again:

<img width="980" alt="TARDIS DAG visualization with backfilled messages in sane order" src="https://user-images.githubusercontent.com/558581/119115196-86356a00-b9ec-11eb-96df-9e7485636be9.png">

Reply:
Resolving in favor of Resolving in favor of https://github.com/matrix-org/matrix-doc/pull/2716#discussion_r655859091